tools/copier_update.py

```diff
@@ -0,0 +1,123 @@
+        return None
+    pr = prs[0]
+    if pr["state"] == "open":
+        return pr["number"]
+    return None
+
+
+def _make_update_dotfiles_pr(org: str, repo: str, branch: str) -> None:
+    subprocess.check_call(
+        ["git", "checkout", "-B", _make_update_dotfiles_branch(branch)]
+    )
+    subprocess.check_call(["git", "add", "."])
+    subprocess.check_call(["git", "commit", "-m", _make_commit_msg(ci_skip=False)])
+    subprocess.check_call(["git", "push", "-f"])
+    if not _get_update_dotfiles_open_pr(org, repo, branch):
+        subprocess.check_call(
+            [
+                "gh",
+                "pr",
+                "create",
+                "--repo",
+                f"{org}/{repo}",
+                "--base",
+                branch,
+                "--title",
+                f"[{branch}] dotfiles update needs manual intervention",
+                "--body",
+                textwrap.dedent(
+                    """\
+                        Dear maintainer,
+
+                        After updating the dotfiles, `pre-commit run -a`
+                        fails in a manner that cannot be resolved automatically.
+
+                        Can you please have a look, fix and merge?
+
+                        Thanks,
+                    """
+                ),
+                "--label",
+                "help wanted",
+            ]
+        )
+
+
+@click.command()
+@click.argument("branch")
+def main(branch: str) -> None:
+    for repo in get_repositories():
+        try:
+            with temporary_clone(repo, branch):
+                print("=" * 10, repo, "=" * 10)
+                # set git user/email
+                subprocess.check_call(
+                    ["git", "config", "user.name", "oca-git-bot"],
+                )
+                subprocess.check_call(
+                    ["git", "config", "user.email", "oca-git-bot@odoo-community.org"],
+                )
+                if not Path(".copier-answers.yml").exists():
+                    print(f"Skipping {repo} because it has no .copier-answers.yml")
+                    continue
+                r = subprocess.call(["copier", "-f", "update"])
+                if r != 0:
+                    print("$" * 10, f"copier update failed on {repo}")
+                    continue
+                # git add updated files so pre-commit run -a will pick them up
+                # (notably newly created .rej files)
+                subprocess.check_call(["git", "add", "."])
+                # run up to 3 pre-commit passes, in case autofixers
+                # (which cause pre-commit to fail when they change files)
+                # resolve issues
+                for _ in range(3):
+                    r = subprocess.call(["pre-commit", "run", "-a"])
+                    # git add, in case pre-commit created new files
+                    subprocess.check_call(["git", "add", "."])
+                    if r == 0:
+                        break
```

Comment:
I think this part should be a little bit improved.

There's 2 things here that are merged but should be IMHO split.

The 1st thing is to reformat dotfiles themselves. This is desired as they will almost 100% of the time will fail (beause `.copier-answers.yml` has to be reformated always).

The 2nd thing is to reformat all the rest of source code.

If you're migrating a module and extract the git history from older branch, and then find in between commits labeled like `[IMP] update dotfiles`, but the real change is not in dotfiles but in source code, that can be a little misleading. Besides, having a clear separation between what's updated by copier and what by pre-commit (in other parts) is also helpful to later debug if there's a problem in the template or in pre-commit configuration.

So, I think it would be better like this (pseudocode):

```py
failed = False

# Update from copier and reformat its changes
_run_copier()
for n in range(3):
    try:
        pre_commit("install")
        git("add", ".")
        git("commit", "-m", "[IMP] update dotfiles")
    except ProcessError:
        if n == 2:
            failed = True
            git("commit", "-m", "[IMP] update dotfiles", "--no-verify")

# Reformat all other code if needed
try:
    pre_commit("run", "-a")
except ProcessError:
    git("add", ".")
    try:
        git("commit", "-m", "[IMP] pre-commit run")
    except ProcessError:
        git("commit", "-m", "[IMP] pre-commit run", "--no-verify")
        failed = True

# Open PR because it failed
if failed:
    _make_update_dotfiles_pr(ORG, repo, branch)
```

Reply:
The more I think to it the more I'm convinced that prettier should not touch files generated by the copier template.
For instance this can cause copier merge conflicts (.rej) as I have had while doing https://github.com/OCA/server-tools/pull/2203, due to the line break added by prettier behind copier's back.

Reply:
I agree

Reply:
It's true, and it might be a valid approach. However, with copier and jinja templating it will always be hard to generate "pretty" and easy to read configuration files, so we can always benefit from pre-commit formatting.

OTOH, it might be that the user edits the file manually, and in a copier update an actual conflict is created, and a `.rej` file is generated. In that case, having pre-commit checking all the files in the repo might be useful... 

Personally, I have gotten pretty used to the flow of (1) `copier update` (2) Run pre-commit on those generated files (3) Fix some conflicts if needed and (4) commit the changes. However, maybe @Yajo can give some more insight on the pros and cons of each option.

Reply:
It's very unlikely that the pre-commit part is guilty of some conflict, as that's supported upstream by copier. I mean that, when doing the previous copy to compare with current state, copier already attempts a 2nd commit if the 1st fails, assuming that in [tasks](https://copier.readthedocs.io/en/stable/configuring/#tasks) you could be doing a `pre-commit install` or equivalent.

That said, another point to take into account is that any copier update can usually mean an update to the pre-commit configuration, which itself could lead to violations of linters/formatters, even in other files. Of course I'm talking of other projects not in OCA, where we freeze that part as much as possible per version.

At the end of the day, it's just a criteria. You can opt for not formatting those files, saving yourself from an extra step, or you can opt for formatting those, producing beautier configs always. I opted for the 2nd, but if you all agree we can use the 1st, it's not a problem for me.